Deprecate `union sock_addr_union` for `struct sockaddr_storage`  and `socklen_param_type` for `socklen_t`

MisterDA opened the pull request:
The two `sock_addr_union` and `socklen_param_type` types were introduced when IPv6 was not widely supported. Modernize the network code by deprecating them and replacing them with `struct sockaddr_storage` and `socklen_t`.
`struct sockaddr_storage` and `socklen_t` were specified in [RFC 2553 Basic Socket Interface Extensions for IPv6](https://datatracker.ietf.org/doc/html/rfc2553) (March 1999). They are also specified by [POSIX 2004 (Issue 6)](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/socket.h.html).
`struct sockaddr_storage` is supported since [Windows XP / Windows Server 2003](https://learn.microsoft.com/en-us/windows/win32/winsock/sockaddr-2). `socklen_t` (defined in [`<ws2tcpip.h>`](https://learn.microsoft.com/en-us/windows/win32/api/ws2tcpip/)) is supported since Windows Vista / Windows Server 2003.

Provide API compatibility with a macro that uses the C11 `_Generic` keyword, and type-punning.

MisterDA committed:
Rename variables and simplify code

MisterDA committed:
Deprecate union sock_addr_union for struct sockaddr API

Use struct sockaddr_storage when the tag of the struct is not known,
and struct sockaddr when it is.
Expand the size of the union to match the size of struct
sockaddr_storage.

Preserve compatibility with union sock_addr_union using C11 _Generic
and C++ function overloading.

MisterDA committed:
Deprecate socklen_param_type for socklen_t

MisterDA committed:
Update changes

MisterDA committed:
Test compatibility with sock_addr_union and socklen_param_type

nojb commented:
`precheck` https://ci.inria.fr/ocaml/job/precheck/1076/

MisterDA commented:
> `precheck` https://ci.inria.fr/ocaml/job/precheck/1076/

precheck is passing, the msvc32 failure is unrelated.

nojb commented on otherlibs/unix/caml/socketaddr.h:
```diff
@@ -71,15 +70,32 @@ extern "C" {
+               struct sockaddr_storage *: (addr),                           \
+               union sock_addr_union *: (struct sockaddr_storage *)(addr)), \
+      (addr_len),                                                           \
+      (close_on_error))
```
I understand that `_Generic` is C11, but since it is the first use in the codebase, are we sure this won't cause any problems on some of the more exotic architectures where OCaml is built?

MisterDA replied:
It's more about the compiler than the architecture. GCC, Clang, and MSVC all support it. `_Generic` is unlikely to be bugged for such simple cases (we haven't mixed in qualifiers, compatible types, typedefs, null pointers…). `_Generic` is typically used to implement type-generic macros for math functions in [`<tgmath.h>`](https://en.cppreference.com/w/c/numeric/tgmath.html) so I'd wager that we won't have surprises with this simple use.

nojb replied:
Yes, I was thinking of some more exotic compilers (other than GCC, Clang, and MSVC) that may be used on some more exotic architectures :) But I don't know if we have many of those nowadays, and anyway, I am not sure that would be an argument against this change, since we do require C11 after all (is this spelled out anywhere?).

MisterDA replied:
The baseline is that we require C11 atomic support. We need GCC 4.9 (2014) for this, fortunately `_Generic` was added at the same time. Clang supports it since 3.0 (2014). MSVC since version 16.7.0 (2019) and we already require a newer version for atomic support. CompCert since version 3.11 (2022). IBM and Intel compilers are now based on Clang, so they should support `_Generic` too. Oracle® Solaris Studio since 12.5 (2016).

- https://gcc.gnu.org/wiki/C11Status
- https://releases.llvm.org/3.0/docs/ClangReleaseNotes.html
- https://clang.llvm.org/c_status.html#c11
- https://learn.microsoft.com/en-us/cpp/c-language/generic-selection?view=msvc-170
- https://compcert.org/

nojb replied:
Thanks for the references. Good enough for me!

nojb commented on otherlibs/unix/socketaddr.c:
```diff
@@ -32,71 +34,68 @@
-      adr->s_unix.sun_family = AF_UNIX;
-      if (len >= sizeof(adr->s_unix.sun_path)) {
+      s_unix->sun_family = AF_UNIX;
+      if (len >= sizeof(((struct sockaddr_un *)0)->sun_path)) {
```
Why not `s_unix->sun_path` here?

MisterDA replied:
Both are equivalent, but I couldn't find a compelling argument for using `((struct sockaddr_un *)0)->` over `s_unix`. I think I like `sizeof` applied on types better. I'll revert this particular change.

nojb reviewed (approved):
Nice cleanup, LGTM (assuming we are OK with the use of `_Generic`).

There is a fair amount of changes in the C code, so a second pair of eyes would be good.

MisterDA commented on Changes:
```diff
@@ -170,6 +170,10 @@ Working version
 
+- #14309: Deprecate union sock_addr_union for struct sockaddr_storage
+  and socklen_param_type for socklen_t.
+  (Antonin Décimo, review by ?)
```
Off-by-one! and I'll add all reviewers.

MisterDA commented:
I think that this patch is correct because if I add enough padding to make `union sock_addr_union` the same size as `struct sockaddr_storage`, then (on my Linux 6.17 x86_64) both trunk `dllunixnat.so` and the same dll with this patch series are bit-for-bit identical. `union sock_addr_union` (112 bytes) is 16 bytes smaller than `struct sockaddr_storage` (128 bytes). I've also inspected the assembly for the fun of it, and the only difference stems from defining the structures on the stack, the offset to address of objects defined after these structs differs by 16 bytes.

```diff
diff --git a/otherlibs/unix/caml/socketaddr.h b/otherlibs/unix/caml/socketaddr.h
index 768669d0cc..8165b8b227 100644
--- a/otherlibs/unix/caml/socketaddr.h
+++ b/otherlibs/unix/caml/socketaddr.h
@@ -52,8 +52,11 @@ union sock_addr_union {
 #ifdef HAS_IPV6
   struct sockaddr_in6 s_inet6;
 #endif
+  uint8_t pudding[sizeof(struct sockaddr_storage)];
 };

+_Static_assert(sizeof(union sock_addr_union) + 16 == sizeof(struct sockaddr_storage), "");
+
 #ifdef HAS_SOCKLEN_T
 typedef socklen_t socklen_param_type;
 #else
```
This should prove that the ABI is respected, and the test should prove that the old API still works.

nojb commented:
Thanks @MisterDA, that's good enough for me and am ready to merge the PR as-is. It is just the introduction of a new language feature (`_Generic`) into the codebase that gives me pause (even if I understand that it should be OK).

@gasche @dra27 @shym do you see any issue with using `_Generic` in the public C headers?

dra27 reviewed (approved):
Looks very nice to me, too, thanks!

I’m happy with the `_Generic` - especially thanks to the test. One question - have you tested this in C++? It wasn’t completely clear to me that this was guaranteed to work?

In passing, the `configure` changes are all in keeping with the discussions on the stalled https://github.com/ocaml/ocaml/pull/10505

MisterDA committed:
Make caml_unix_alloc_sockaddr take a struct sockaddr *

rather than a struct sockaddr_storage *, as its an _in_ parameter and
its type is known, thus its size.

MisterDA commented:
> One question - have you tested this in C++? It wasn’t completely clear to me that this was guaranteed to work?

A good catch indeed! I had not, and it did not. But wait to hear about this trick, OCaml programmers hate it, it's function overloading!

I think the code is correct now. The C++ with bytecode test would complain that the external functions were missing, so I kept only the native test.

MisterDA committed:
Use closesocket instead of close when applied to a socket on Windows

MisterDA commented:
I've added two more changes to this series since the last review:
- Make `caml_unix_alloc_sockaddr` take a `struct sockaddr *` rather than a `struct sockaddr_storage *`, as its an _in_ parameter and its type is known, thus its size. I think this makes more sense for external users rather than requiring the struct to be wider.
- Use `closesocket` instead of `close` when applied to a socket on Windows in the error path of `caml_unix_alloc_sockaddr`.

dra27 commented:
> I think the code is correct now. The C++ with bytecode test would complain that the external functions were missing, so I kept only the native test.

I don’t quite follow this - what’s the issue? I can see it was the case before, but is the test skipped on MSVC just to avoid the multiple compiler syntaxes? (rather than anything more fundamental)

MisterDA commented:
> > I think the code is correct now. The C++ with bytecode test would complain that the external functions were missing, so I kept only the native test.
> 
> I don’t quite follow this - what’s the issue?

If I add a bytecode test:
```
 {
   setup-ocamlc.byte-build-env;
   script = "${cc} -xc++ -std=c++11 \
     -I ${ocamlsrcdir}/runtime \
     -I ${ocamlsrcdir}/otherlibs/unix \
     -o ${test_build_directory}/sockaddr_cxx_aux.o \
     -c ${test_source_directory}/sockaddr_cxx_aux.cpp";
   script;
   all_modules = "sockaddr_cxx_aux.o sockaddr_cxx.ml";
   ocamlc.byte;
   output = "${test_build_directory}/program-output";
   stdout = "${output}";
   run;
   check-program-output;
 }
 ```
 
 I get this error:
 ```
 > Running action 1/1 (ocamlc.byte)
> Compiling program /Users/antonin/Tarides/ocaml/trunk/testsuite/_ocamltest/tests/lib-unix/unix-sockaddr/sockaddr_cxx/ocamlc.byte/sockaddr_cxx.byte from modules  sockaddr_cxx_aux.o sockaddr_cxx.ml
> Commandline: /Users/antonin/Tarides/ocaml/trunk/runtime/ocamlrun /Users/antonin/Tarides/ocaml/trunk/ocamlc -use-runtime /Users/antonin/Tarides/ocaml/trunk/runtime/ocamlrun -I /Users/antonin/Tarides/ocaml/trunk/runtime -nostdlib -I /Users/antonin/Tarides/ocaml/trunk/stdlib -I /Users/antonin/Tarides/ocaml/trunk/otherlibs/unix unix.cma -o /Users/antonin/Tarides/ocaml/trunk/testsuite/_ocamltest/tests/lib-unix/unix-sockaddr/sockaddr_cxx/ocamlc.byte/sockaddr_cxx.byte sockaddr_cxx_aux.o sockaddr_cxx.ml
>   Redirecting stdout to /Users/antonin/Tarides/ocaml/trunk/testsuite/_ocamltest/tests/lib-unix/unix-sockaddr/sockaddr_cxx/ocamlc.byte/ocamlc.byte.output
>   Redirecting stderr to /Users/antonin/Tarides/ocaml/trunk/testsuite/_ocamltest/tests/lib-unix/unix-sockaddr/sockaddr_cxx/ocamlc.byte/ocamlc.byte.output
> ### begin stdout ###
> File "sockaddr_cxx.ml", line 1:
> Error: Error while linking sockaddr_cxx.cmo:
>        The external function stubs_old is not available
> ### end stdout ###
> Action 1/1 (ocamlc.byte) => failed (Compiling program /Users/antonin/Tarides/ocaml/trunk/testsuite/_ocamltest/tests/lib-unix/unix-sockaddr/sockaddr_cxx/ocamlc.byte/sockaddr_cxx.byte from modules  sockaddr_cxx_aux.o sockaddr_cxx.ml: command
> /Users/antonin/Tarides/ocaml/trunk/runtime/ocamlrun /Users/antonin/Tarides/ocaml/trunk/ocamlc -use-runtime /Users/antonin/Tarides/ocaml/trunk/runtime/ocamlrun -I /Users/antonin/Tarides/ocaml/trunk/runtime  -nostdlib -I /Users/antonin/Tarides/ocaml/trunk/stdlib -I /Users/antonin/Tarides/ocaml/trunk/otherlibs/unix  unix.cma    -o /Users/antonin/Tarides/ocaml/trunk/testsuite/_ocamltest/tests/lib-unix/unix-sockaddr/sockaddr_cxx/ocamlc.byte/sockaddr_cxx.byte   sockaddr_cxx_aux.o sockaddr_cxx.ml
> failed with exit code 2)
```

The fix could be to generate a complete executable, but I'm not sure where to go from here. As the native test worked, I disabled the bytecode.

> I can see it was the case before, but is the test skipped on MSVC just to avoid the multiple compiler syntaxes? (rather than anything more fundamental)

Yes, exactly. If I could get the ccomp_type I could add both, maybe?

shym commented on otherlibs/unix/caml/socketaddr.h:
```diff
@@ -71,15 +70,34 @@ extern "C" {
+  caml_unix_get_sockaddr(                                                   \
+    (vaddr),                                                                \
+    _Generic((addr),                                                        \
+             struct sockaddr_storage *: (addr),                             \
```
Could, or should, this be `default: ...` instead? I think the error messages (at least with gcc) will make more sense to the end user who probably doesn’t care that much about the fact that a `_Generic` didn’t find a matching type.
(And similar remark for the other `_Generic`, of course)

MisterDA replied:
Good point, thanks. I've applied this.

shym commented on otherlibs/unix/caml/socketaddr.h:
```diff
@@ -92,6 +110,25 @@ extern value caml_unix_alloc_inet6_addr (struct in6_addr * inaddr);
+   struct sockaddr_storage. C++ doesn't support C11 _Generic. */
+static inline void
+caml_unix_get_sockaddr(value vaddr, union sock_addr_union * addr /*out*/,
+                       socklen_param_type * addr_len /*out*/) {
+  caml_unix_get_sockaddr(vaddr,
+                         reinterpret_cast<struct sockaddr_storage *>(addr),
+                         addr_len);
+}
+static inline value
+caml_unix_alloc_sockaddr(union sock_addr_union * addr /*out*/,
+                         socklen_param_type addr_len /*out*/,
+                         int close_on_error) {
+  return caml_unix_alloc_sockaddr(
+    reinterpret_cast<struct sockaddr *>(addr),
+    addr_len,
+    close_on_error);
+}
```
I think it would make it more legible to move that in an `#else` just after the C version, instead of hiding it with the `external "C"` block end.

How about saying in the comment: `C++ doesn't support C11 _Generic, so use function overloading.`?

MisterDA replied:
Good point too, thanks! I've applied this.

shym reviewed (approved):
I have no particular insight about possible issues with `_Generic`, unfortunately.

I would probably have added a `struct sockaddr_storage` field in `union sock_addr_union` and used that field instead of explicit casts, even though I’m never sure what’s the most robust technique for type punning.

MisterDA commented:
> The fix could be to generate a complete executable, but I'm not sure where to go from here. As the native test worked, I disabled the bytecode.

I've fixed and re-enabled the bytecode test for C++.
I'm not too much interested in testing support for MSVC and C++.

I've applied the changes suggested by @shym, thanks for the review.

> I would probably have added a `struct sockaddr_storage` field in `union sock_addr_union` and used that field instead of explicit casts, even though I’m never sure what’s the most robust technique for type punning.

I'd like the calling code to use the standard APIs, I'm not sure how adding a new field would enable that.

MisterDA commented:
>> I would probably have added a `struct sockaddr_storage` field in `union sock_addr_union` and used that field instead of explicit casts, even though I’m never sure what’s the most robust technique for type punning.
>
> I'd like the calling code to use the standard APIs, I'm not sure how adding a new field would enable that.

We clarified this offline, and I've pushed a new-new-new revision that, I hope, addresses all the concerns.

shym commented on otherlibs/unix/strofaddr.c:
```diff
@@ -28,22 +28,22 @@ CAMLprim value caml_unix_string_of_inet_addr(value a)
-    memset(&sa.s_inet6, 0, sizeof(struct sockaddr_in6));
-    sa.s_inet6.sin6_family = AF_INET6;
-    sa.s_inet6.sin6_addr = GET_INET6_ADDR(a);
+    memset(&addr, 0, sizeof(struct sockaddr_storage));
```
```suggestion
    memset(&addr, 0, sizeof(struct sockaddr_storage));
```
This line and the one in the other branch can now be factorised.

MisterDA replied:
Done! Note that the Windows code path can be completely eliminated by switching to `inet_ntop`/`InetNtop`. I'll do that in a follow-up.

shym reviewed (approved):
LGTM, thank you!

MisterDA committed:
strofaddr: minor refactor in Windows code path

MisterDA commented:
merge-me? 👼
